### PR TITLE
Add an "is_ignored" field to JSON output

### PR DESCRIPTION
**Pull request by underyx**
This makes it possible for semgrep-agent to tell when a finding has been
ignored with a nosem comment.

Changes were tested with the included test named
`test_nosem_rule__with_disable_nosem`,
as well as reading through the updated snapshots.

### CODE REVIEW

**Comment by underyx** [comment]
Force-pushed a changelog entry.

**Comment by mschwager** [comment]
Looks like this is related to https://github.com/returntocorp/semgrep/issues/2033. Although, I don't understand: why does semgrep-agent need to know if a finding has been ignored? Seems like an interesting data point, but shouldn't Semgrep just not return the finding if it's been ignored?

Side note: this would be much easier to review if the refactor and functionality changes were done separately :)

**Comment by mschwager** [comment]
Actually, taking a step back... I don't understand what the original bug here was. Could you briefly describe the issue, and how this solves it?

**Comment by underyx** [comment]
@mschwager we want to be able to show and crunch data about ignored issues in Semgrep Community. This will be able to flip issues from `NEW` to `IGNORED` status in semgrep.dev's db instead of `FIXED` being the only resolution option.

**Comment by underyx** [comment]
I re-ran the benchmark job, not sure what could make semgrep +5% slower here :/

Is the baseline from the current develop version?

**Comment by underyx** [comment]
Seems like the benchmark re-run fixed it!

File: semgrep/semgrep/semgrep_main.py [hunk 1]
```diff
@@ -231,12 +233,18 @@ def main(
 
     output_handler.handle_semgrep_errors(semgrep_errors)
 
+    rule_matches_by_rule = {
```
Comment (brendongo) [inline semgrep/semgrep/semgrep_main.py:236]: Should this only run as an else to the conditional below?

**Review by underyx — COMMENTED** [review]
@brendongo re: your deleted comment:

> Could go either way. I ended up including `is_ignored: false` on each run just to make the JSON output schema more consistent.

**Comment by brendongo** [comment]
> @brendongo re: your deleted comment:
> 
> > Could go either way. I ended up including `is_ignored: false` on each run just to make the JSON output schema more consistent.

Yeah it was more clear on second reading. :D

**Review by brendongo — APPROVED** [review]
LGTM. If anything I would change is_ignored to something like has_no_sem or something to make it more obvious what it does but that seems like a nit.

**Comment by mschwager** [comment]
Discussed out-of-band - not related to https://github.com/returntocorp/semgrep/issues/2033. LGTM :+1: although yeah I agree that naming tying this closer to `nosem` would be helpful

**Comment by underyx** [comment]
`ignored_with_nosem`?

**Comment by underyx** [comment]
Though what if we add more methods of ignoring, for instance for ignoring a rule ID for a whole file with a marker on line 1? It's conceivable that new ignore methods wouldn't be named nosem anymore 🤔

**Comment by underyx** [comment]
My issue with just `has_nosem` is that without the word "ignore" someone looking at the JSON output who isn't aware of `// nosem` functionality would have zero idea what this is supposed to mean; they might as well think it's about people's noses or something.

**Comment by brendongo** [comment]
> Though what if we add more methods of ignoring, for instance for ignoring a rule ID for a whole file with a marker on line 1? It's conceivable that new ignore methods wouldn't be named nosem anymore 🤔

hmm makes sense. I can imagine we might want to know how a finding was ignored if there are many ways to ignore. I'm fine with whatever you choose here @underyx can always change this later.

**Comment by underyx** [comment]
So I'll merge as is. I expect we'll be the only users of this field so changing it later won't be likely to hurt anyone.

**Pull request merged**